prime-router/docs/proposals/lineage/README.md

```diff
@@ -0,0 +1,123 @@
+
+An **OrganizationService** is an endpoint to which the Hub can send a specific kind of data to a specific place, so that the Organiation can use the data.  For example, an OrganizationService might be an sftp site where a local health department wants hl7 files uploaded every 4 hours.
+
+Likewise, an **OrganizationClient** is an endpoint that sends data to the Hub API.   For example, SimpleReport has an OrganizationClient that can send .csv data to the hub, as often as they like.
+
+
+## Assumptions
+
+- TASKs create zero or more REPORTs which have one or more ITEMs
+- ITEMs are like quarks.They only exist inside of a REPORT.  Its fine and quite common to have a report with one item.
+- We must track lineage per-item, however, people think in terms of REPORTs.  Hence we need to track both.
+- Tracking of sending- and receiving- Organizations is crucial, to know who gets permissions to see the data.
+- Many:many -- A given ACTION could take in many reports, and produce many reports.  Need to allow for both splitting and merging.
+- Many:many -- An ITEM can be the result of multiple parent items, and can result in multiple child ITEMs.
+- A few schema best practices:  Every table must have a single col PK (either a serial or a UUID), and a created_date, Normalize independent data.   Use singular nouns.
+-- Receipt of a dataset implies the right to see the lineage of that dataset.
+-- Sending a dataset implies the right to see the lineage of that dataset. 
+
+## Proposed Design
+
+Please [pull up the ER diagram](./lineage-er-diagram.jpeg) for the detailed proposed schema.
+
+(Note: [Original lucid chart diagram is here](https://lucid.app/lucidchart/7b81cccb-91a0-44b3-b17f-90b9b0c8304e/edit?beaconFlowId=332110D5160D6847&page=0_0#))
+
+### Notes on the Proposed Schema
+
+- The separation of Task from Report allows us to have Tasks that don't necessarily create Reports.
+- Addition of the `TASK.task_result` fills a gap where we're not tracking the error messages sent back to users.
+- Apparently I can't name the REPORT table 'REPORT', due to a limitation in Flyway/jooq.  Hence the weird name `REPORT_FILE`.   I also thought about `REPORT_IN`, `REPORT_DATA`, `REPORT_REC`.  Haven't sunk as low as `REPURT`, but I'm on my way there...
+- Many values in REPORT_FILE will be empty depending on the task creating the report.  For example, all the sending_* fields are only filled in for an initially received report.   All the receiving_* fields are only filled in for reports sent externally.
+- might need a report_status in the future.  For example, `wiped` or `archived`, to explain why a blob might have disappeared.
+- The primary key to use for ITEM is tricky.  The sender gives us the `tracking_element`, but we have little control over how robust that is.  Because of this, all queries on `tracking_element` must also filter on report_id.
```

Comment:
Agreed. Maybe we should manufacture a tracking_element ID if one is not provided by the sender?

Reply:
Not sure I'm a fan of that.   We need a strong way to link data back to what the sender sent, when there are questions or errors.   I think it should be an error if sender fails to send a proper tracking_element_id.